File: ruby/tabulation/canSum.rb
def can_sum(targetSum, numbers)
    table = Array.new(targetSum+1,false)
    table[0] = true

    (targetSum+1).times do |i|
        if table[i] === true 
            numbers.each do |number|
                table[i + number] = true
            end
        end
    end
    table[targetSum]
end

p can_sum(14,[9,9,5])

File: javascript/tabulation/canSum.js
const canSum = (targetSum, numbers) => {
  const table = Array(targetSum + 1).fill(false);
  table[0] = true;

  for (let i = 0; i <= targetSum; i++) {
    if (table[i] === true) {
      for (let number of numbers) {
        table[i + number] = true;
      }
    }
  }

  return table[targetSum];
};

// time O(mn)
// space O(m)

console.log(canSum(7, [5, 8, 3]));
